revamp Travis-CI builds

stonerl opened the pull request:
Travis-CI builds are completely rewritten and don't require sudo anymore.

macOS: We build on 10.14(clang 7), 10.13(clang 5) & 10.12(clang3.9)
Linux: We build on 18.04 (clang 7 & gcc 7.4), 16.04 (gcc 5.4) & 14.04 (clang 3.4 & gcc 4.8)

On macOS all builds use ccache on Linux only the gcc Release builds. Since clang compiles
faster it is not as big of a problem and jobs are within the travis time-limit.

Regression suite is run on Linux jobs only if the compile-time didn't exceed 33 mins, to
avoid hitting the time-limit of 50 mins while executing the tests. This mainly affects
gcc debug builds.

For xvfb we utilize the new service stanza - only available on 18.04 and 16.04, though.
For 14.04 we use the old fashion way.

Only pushes to master and PR trigger builds.

stonerl committed:
revamp Travis-CI builds

Travis-CI builds are completely rewritten and don't require sudo anymore.

macOS: We build on 10.14(clang 7), 10.13(clang 5) & 10.12(clang3.9)
Linux: We build on 18.04 (clang 7 & gcc 7.4), 16.04 (gcc 5.4) & 14.04 (clang 3.4 & gcc 4.8)

On macOS all builds use ccache on Linux only the gcc Release builds. Since clang compiles
faster it is not as big of a problem and jobs are within the travis time-limit.

Regression suite is run on Linux jobs only if the compile-time didn't exceed 33 mins, to
avoid hitting the time-limit of 50 mins while executing the tests. This mainly affects
gcc debug builds.

For xvfb we utilize the new service stanza - only available on 18.04 and 16.04, though.
For 14.04 we use the old fashion way.

Only pushes to master and PR trigger builds.

gunchleoc commented:
Looking at this: https://travis-ci.org/widelands/widelands/builds/587774617

I can't see the compiler versions for each build any more, except for Mac builds.

stonerl committed:
add name to build-matrix

The build jobs are named now. Containing OS and compiler-version.

stonerl committed:
add build type to name

gunchleoc commented:
Yep, that fixed it. Let's wait for Travis to finish and then merge if it looks OK.

stonerl commented:
Added DEBUG/RELEASE as a prefix.

stonerl committed:
Make use of stages

We use builds stages now. First we run a code check and test the documentation. If nothing fails we the can go on and start compiling. This saves roughly 2 minutes on the debug builds which should help us with timeouts. Furthermore, we don't waste time when the code isn't clean.

stonerl committed:
change whitelist

remove travis-with-caches whitelist

stonerl commented:
Should be better now. Compiling doesn't start when the code isn't clean or building the documentation fails. Removing the code check from all debug builds saves roughly 2-3 minutes on each job.

gunchleoc reviewed (approved):
Perfect :)

I think we should switch off codecheck for the debug builds now rather than skipping the test suite. That requires some CMake change though, so let's do that in a follow-up branch.
